lib/host_runner.rb

```diff
@@ -35,10 +47,60 @@ def start
       raise LockError.new("Host '#{@host_name}' already locked")
     end
 
+    connect
     @ip
   end
 
+  def cleanup
+    run_command "snapper", "create", "-c", "number", "--pre-number", @base_snapshot_id.to_s,
+      "--description", "pennyworth_snapshot"
+    run_command "snapper", "undochange", "#{@base_snapshot_id}..0"
+    run_command "bash", "-c", "reboot &"
+  end
+
   def stop
+    if @connected && !ENV["SKIP_CLEANUP"]
+      cleanup
+    end
     @locker.release_lock(@host_name)
   end
+
+  private
+
+  # Makes sure the we can connect to the remote system as root (without a
+  # password or passphrase) and that all required binaries are available.
+  def connect
+    begin
+      Cheetah.run "ssh", "-q", "-o", "BatchMode=yes", "root@#{@ip}"
+    rescue Cheetah::ExecutionFailed
+      raise SshConnectionFailed.new(
+        "Could not establish SSH connection to host '#{@ip}'. Please make sure that " \
+        "you can connect non-interactively as root, e.g. using ssh-agent.\n\n" \
+        "To copy your default ssh key to the machine run:\n" \
+        "ssh-copy-id root@#{@ip}"
+      )
+    end
+
+    begin
+      Cheetah.run "snapper", "--help"
+    rescue Cheetah::ExecutionFailed
+      raise CommandNotFoundError.new(
+        "Snapper needs to be installed on the test system '#{@ip}' for the automatic cleanup."
+      )
+    end
+    @connected = true
+  end
+
+  def run_command(*cmd)
```

Comment:
I think I would extract the Cheetah calls to execute a remote command to a separate class `RemoteCommand` and then dry up the code in `VM` and here by using this class. We would need a few parameters or options to accommodate the different invocations, but there is quite some duplication we could get rid of by that.

This also would make the code easier to read, because it would make it explicit in the calling code, what is supposed to happen. This is not that easy to find out from reading the different Cheetah calls.